lib/ruby/truffle/truffle/objspace.rb

```diff
@@ -0,0 +1,271 @@
+# Copyright (c) 2015 Oracle and/or its affiliates. All rights reserved. This
+# code is released under a tri EPL/GPL/LGPL license. You can use it,
+# redistribute it and/or modify it under the terms of the:
+#
+# Eclipse Public License version 1.0
+# GNU General Public License version 2
+# GNU Lesser General Public License version 2.1
+
+require 'json'
+require 'tempfile'
+require 'weakref'
+
+module ObjectSpace
+
+  def count_nodes(nodes = {})
+    ObjectSpace.each_object(Module) do |mod|
+      mod.methods(false).each do |name|
+        count_nodes_method mod.method(name), nodes
+      end
+
+      mod.private_methods(false).each do |name|
+        count_nodes_method mod.method(name), nodes
+      end
+    end
+
+    ObjectSpace.each_object(Proc) do |proc|
+      count_nodes_method proc, nodes
+    end
+
+    ObjectSpace.each_object(Method) do |method|
+      count_nodes_method method, nodes
+    end
+
+    ObjectSpace.each_object(UnboundMethod) do |umethod|
+      count_nodes_method umethod, nodes
+    end
+
+    nodes
+  end
+
+  module_function :count_nodes
+
+  class << self
+
+    def count_nodes_method(method, nodes)
+      count_nodes_tree Truffle::Primitive.ast(method), nodes
+    end
+
+    private :count_nodes_method
```

Comment:
For style, usually the `private` follows the `end` or then the scope/frame-level `private` is used only once (better in this case I think).
A third alternative from Ruby 2.2 is `private def count_nodes_method(method, nodes)` but might be strange for indent reasons.